stumpy/stump_tiles.py

```diff
@@ -0,0 +1,794 @@
+        The (top-k) matrix profile
+
+    indices : numpy.ndarray
+        The (top-k) matrix profile indices
+
+    left profile : numpy.ndarray
+        The (top-1) left matrix profile
+
+    left indices : numpy.ndarray
+        The (top-1) left matrix profile indices
+
+    right profile : numpy.ndarray
+        The (top-1) right matrix profile
+
+    right indices : numpy.ndarray
+        The (top-1) right matrix profile indices
+
+    Notes
+    -----
+    `DOI: 10.1007/s10115-017-1138-x \
+    <https://www.cs.ucr.edu/~eamonn/ten_quadrillion.pdf>`__
+
+    See Section 4.5
+
+    The above reference outlines a general approach for traversing the distance
+    matrix in a diagonal fashion rather than in a row-wise fashion.
+
+    `DOI: 10.1145/3357223.3362721 \
+    <https://www.cs.ucr.edu/~eamonn/public/GPU_Matrix_profile_VLDB_30DraftOnly.pdf>`__
+
+    See Section 3.1 and Section 3.3
+
+    The above reference outlines the use of the Pearson correlation via Welford's
+    centered sum-of-products along each diagonal of the distance matrix in place of the
+    sliding window dot product found in the original STOMP method.
+
+    `DOI: 10.1109/ICDM.2016.0085 \
+    <https://www.cs.ucr.edu/~eamonn/STOMP_GPU_final_submission_camera_ready.pdf>`__
+
+    See Table II
+
+    Timeseries, T_A, will be annotated with the distance location
+    (or index) of all its subsequences in another times series, T_B.
+
+    Return: For every subsequence, Q, in T_A, you will get a distance
+    and index for the closest subsequence in T_B. Thus, the array
+    returned will have length T_A.shape[0]-m+1. Additionally, the
+    left and right matrix profiles are also returned.
+
+    Note: Unlike in the Table II where T_A.shape is expected to be equal
+    to T_B.shape, this implementation is generalized so that the shapes of
+    T_A and T_B can be different. In the case where T_A.shape == T_B.shape,
+    then our algorithm reduces down to the same algorithm found in Table II.
+
+    Additionally, unlike STAMP where the exclusion zone is m/2, the default
+    exclusion zone for STOMP is m/4 (See Definition 3 and Figure 3).
+
+    For self-joins, set `ignore_trivial = True` in order to avoid the
+    trivial match.
+
+    Note that left and right matrix profiles are only available for self-joins.
+    """
+    n_A = T_A.shape[0]
+    n_B = T_B.shape[0]
+    l = n_A - m + 1
+    n_threads = numba.config.NUMBA_NUM_THREADS
+
+    ρ = np.full((n_threads, l, k), np.NINF, dtype=np.float64)
+    I = np.full((n_threads, l, k), -1, dtype=np.int64)
+
+    ρL = np.full((n_threads, l), np.NINF, dtype=np.float64)
+    IL = np.full((n_threads, l), -1, dtype=np.int64)
+
+    ρR = np.full((n_threads, l), np.NINF, dtype=np.float64)
+    IR = np.full((n_threads, l), -1, dtype=np.int64)
+
+    tile_length = config.STUMPY_MAX_TILE_LENGTH
+    tiles = core._get_tiles(m, n_A, n_B, diags, tile_length)
+    tiles_ranges = core._get_array_ranges(tiles[:, 6], n_threads, False)
```

Comment:
This can become inefficient when length of time series `T` is small. Because, based on my understanding, `len(tiles[:, 6])` is the number of tiles. So, with `STUMPY_MAX_TILE_LENGTH = 4096`, we have:

* if `len(T) = 1000`, we have one tile only!
* if `len(T) = 10000`, we have 9 tiles only (in fact, I think four of them are actually large-size tiles, and five of them are small-size tiles. this is because 10000 is not exactly dividable by 4096 and we will have tiles with different sizes.)

Reply:
@seanlaw 
btw, I think the example below is interesting:
```
n = 10000
m = 50

seed = 0
np.random.seed(seed)
T_A = np.random.rand(n)
T_B = T_A.copy()

n_A = len(T_A)
n_B = len(T_B)

diags = np.arange(0, n_A - m + 1)
tile_length = config.STUMPY_MAX_TILE_LENGTH # 4096
n_threads = numba.config.NUMBA_NUM_THREADS # 8

tiles = core._get_tiles(m, n_A, n_B, diags, tile_length)
tiles_ranges = core._get_array_ranges(tiles[:, 6], n_threads, False) 
```

**Let us take a look at `tiles[:, 6]`:**
```
array([ 8390656, 16777216,  7204864,        0,  8390656,  7204864,
              0,        0,  1547920], dtype=int64)
```

**And, now `tiles_ranges`**
```
>>> array([[0, 1],
       [1, 2],
       [2, 9],
       [9, 9],
       [9, 9],
       [9, 9],
       [9, 9],
       [9, 9]], dtype=int64)
```

This means that only the first two threads are useful!

Reply:
A couple of points:

1. Currently, the way that `stump_tiles` is implemented, it certainly targets larger length time series and corresponds to the "switch" that we'll need to implement later to automatically choose a heuristic for when `_stump` vs `_stump_tiles` is used
2. `config.STUMPY_MAX_TILE_LENGTH` is not at all optimized and is something we need to do. In reality, we should test for much smaller tile sizes like `4 x 4` and go up from there to see its affects on performance (I wouldn't optimize for `k` at all for now and only for `n` and `config.STUMPY_MAX_TILE_LENGTH`).

Reply:
I see! Okay...I will ignore this then :)